### DISCUSSION TITLE
Can't bundle jspdf into custom endpoint

### DISCUSSION THREAD

**Discussion by AlphaMoonbaseBerlin** [directus/directus#10893] [Archive]
I try to implement jspdf into a custom-endpoint.
When trying to build the extension, rollup throws the following error:

> Error: Invalid value for option "output.file" - when building multiple chunks, the "output.dir" option must be used, not "output.file". To inline dynamic imports, set the "inlineDynamicImports" option.

This seems to indikate that the rollup-implementation as set up to package everything into a single file.

I digged arround a little bit into the script and found that there is the chance to create a _extensions.config.js_ to pass data to rollup. But trying to force it to either use dir-mode instead of file-mode or allowing inlineDynamicImports does not seem to work as intendet as in it does not have any effect. But maybe I'm missing some part here in how to setup the _extensions.config.js_ correctly to pass that information:
```
export default {
    output: {
        dir: './dist'
      },
  }
```

Maybe this is not per se a bug but a feature-request or asking for en enhancment of the docs in regards to the rollup-config.

### To Reproduce

npm install jspdf

index.js:
`import { jsPDF } from "jspdf";`

npm run build

### Errors Shown

```
✖ Failed

Error: Invalid value for option "output.file" - when building multiple chunks, the "output.dir" option must be used, not "output.file". To inline dynamic imports, set the "inlineDynamicImports" option.
[Error] https://rollupjs.org/guide/en/#outputdir
Error: Invalid value for option "output.file" - when building multiple chunks, the "output.dir" option must be used, not "output.file". To inline dynamic imports, set the "inlineDynamicImports" option.
```

### What version of Node.js are you using?

14.7.0

### What operating system are you using?

Windows 10

**Comment by rijkvanzanten** [comment]
Looks like jsPDF is doing it's own dynamic import magic, which isn't compatible with the current rollup configuration. Any thoughts @nickrum

**Reply by nickrum** [reply]
We should definitely enable rollup's `inlineDynamicImports` option, so we will always generate a single source file. Though, I'm not sure if that will solve the issue. `jspdf` exports a node-only cjs entrypoint and a browser-compatible ESM entrypoint. Rollup prefers the ESM one, while the extensions-sdk would ideally use the node-only one when bundling API extensions.

> Maybe this is not per se a bug but a feature-request or asking for en enhancment of the docs in regards to the rollup-config.

Right now the config file only supports adding additional rollup plugins as described [here](https://docs.directus.io/extensions/creating-extensions/#configuring-the-cli). It would be nice to extend this in the future to allow setting all sorts of rollup options that might be useful.

**Reply by AlphaMoonbaseBerlin** [reply]
> so we will always generate a single source file.

Not knowing the internals, why would it be a problem to have more then one file?

**Reply by nickrum** [reply]
It's very important for App extensions, because we only serve a single source file per extension type. But even for API extensions, it's a good idea to have a single file because the new auto reloading functionality will only invalidate the extension's entrypojnt when reloading.

**Comment by sebastianomorando** [comment]
Hi! I had the same problem and I resolved importing jspdf this way: `import jsPDF from 'jspdf/dist/jspdf.umd.js'`

**Reply by AlphaMoonbaseBerlin** [reply]
Awsome, thats super handy to know !
Another solution would also be to use require and npm install the package into directus, but thats kinda unhandy atm.
